handout/testing/unit_testing_services.md

```diff
@@ -0,0 +1,138 @@
+    import {SearchWiki} './wikisearch';
+    
+    describe("Testing the wikipedia search service", () => {
+      beforeEachProviders(() => {
+      	return [
+      		MockBackend,
+          	BaseRequestOptions,
+          	SearchWiki,
+          	provide(
+              Http, {
+                useFactory: (
+                  mockbackend: ConnectionBackend, defaultOptions: BaseRequestOptions
+                ) => {
+      				return new Http(mockbackend, defaultOptions)
+    			},
+                deps: [MockBackend, BaseRequestOptions]
+              }
+    	   )
+    	]
+      });
+    });
+```
+As you can see, we have quite a few imports here to setup our mocked HTTP module. We use beforeEachProvider to include MockBackend to create our mocked HTTP module. In order to properly setup our mocked module we also include `BaseRequestOptions` and `ConnectionBackend`. We create an instance of `Http` that our service will get by using a factory. To create this instance the `Http` class take two parameters - the first is the backend to use (it has the type `ConnectionBackend`), and the second is the default options to use (of type `RequestOptions`). Since we are planning to use a mock backend, we feed in `MockBackend` to the first parameter. The second parameter, default options, doesn't have any bearing on the mocking aspects we plan on using, so we use `BaseRequestOptions`, which is the default implementation. 
+
+Right now this test doesn't do anything aside from setting up a mockable HTTP module - lets take a look at some actual tests. 
+
+##Testing HTTP Requests
+Now that we have setup our service and created a mocked version of the HTTP module we are ready to start testing. The basic strategy for testing this service is to verify the contents of the request being made (correct URL) and ensuring that the data we mock into the service is returned correctly by the right method (calling search will return the data we expect it to). 
+
+###Asynchronous Testing
+Services, by their nature, perform asynchronous tasks. When we make an HTTP request we do so in an asynchronous manner so as not to block the rest of the application from carrying out its operations. We you recall, in a previous section we looked a bit at testing components asynchronously. Fortunately a lot of this knowledge carries over into testing services asynchronously. In the `angular2/testing module` there are a few functions that allow us to fake asynchronous behaviour. This is done by wrapping the entire test in a call to `fakeAsync`, and then using calls to `tick` to move the test along. FakeAsync essentially puts our test into a zone, and then listens for any asynchronous operations like setTimeouts, Promises, callbacks, etc. Those functions are not actually called asynchronously, instead Angular will rely on us calling tick, which will call those functions immediately, wait for their execution to finish, and then proceed, thus simulating time elapsing. 
```

Comment:
"If you recall, in a previous section..."

"we looked briefly at testing components asynchronously"